internal/provider/resource_private_key_test.go

```diff
@@ -15,36 +15,45 @@ func TestPrivateKeyRSA(t *testing.T) {
 		Steps: []r.TestStep{
 			{
 				Config: `
-                    resource "tls_private_key" "test" {
-                        algorithm = "RSA"
-                    }
-                    output "private_key_pem" {
-                        value = "${tls_private_key.test.private_key_pem}"
-                    }
-                    output "public_key_pem" {
-                        value = "${tls_private_key.test.public_key_pem}"
-                    }
-                    output "public_key_openssh" {
-                        value = "${tls_private_key.test.public_key_openssh}"
-                    }
-                    output "public_key_fingerprint_md5" {
-                        value = "${tls_private_key.test.public_key_fingerprint_md5}"
-                    }
-                `,
+					resource "tls_private_key" "test" {
+						algorithm = "RSA"
+					}
+					output "private_key_pem" {
```

Comment:
Ah, this resource is using legacy style testing -- it might be worth updating it to use the helper/resource TestCheckFunc, such as [`TestMatchResourceAttr()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#TestMatchResourceAttr) directly against the resource attributes rather than creating outputs and referencing the state directly.

For example:

```go
Check: resource.ComposeAggregateTestCheckFunc(
  resource.TestMatchResourceAttr("tls_private_key.test", "private_key_pem", regexp.MustCompile(`^-----BEGIN OPENSSH PRIVATE KEY-----`)),
  // ... other TestCheckFunc ...
),
```

Refer to the [Terraform website](https://www.terraform.io/plugin/sdkv2/testing/acceptance-tests/teststep#builtin-check-functions) documentation for more details.

Reply:
Will try to refactor.

Reply:
I didn't want to leave this without giving it a go:

```
{
Config: `
	resource "tls_private_key" "test" {
		algorithm = "RSA"
		rsa_bits = 4096
	}
	output "key_pem" {
		value = "${tls_private_key.test.private_key_pem}"
		sensitive = true
	}
`,
Check: r.ComposeAggregateTestCheckFunc(
	r.TestMatchResourceAttr("tls_private_key.test", "private_key_pem", regexp.MustCompile(`^-----BEGIN RSA PRIVATE KEY-----`)),
	func(s *terraform.State) error {
		pem := s.RootModule().Resources["tls_private_key.test"].Primary.Attributes["private_key_pem"]
		if len(pem) < 1700 {
			return fmt.Errorf("key PEM looks too short for a 4096-bit key (got %v characters)", len(pem))
		}
		return nil
	},
),
},
```

It definitely improves the length of tests, but I did find an example of function we might want to add to the `resource` package: something like a `TestResourceAttrLength()`. We could offer one that checks "string is between a range min/max", and another one "string length is".

Reply:
I just created this PR https://github.com/hashicorp/terraform-plugin-sdk/pull/893 that would help avoiding having to write that `func` in the example above to check the PEM length.